src/librustc_session/session.rs

```diff
@@ -1021,6 +1021,32 @@ impl Session {
         || self.opts.debugging_opts.sanitizer.intersects(SanitizerSet::ADDRESS | SanitizerSet::MEMORY)
     }
 
+    pub fn link_dead_code(&self) -> bool {
+        match self.opts.cg.link_dead_code {
+            Some(explicitly_set) => explicitly_set,
+            None => {
+                self.opts.debugging_opts.instrument_coverage
+                    && !self.target.target.options.is_like_msvc
+                // Issue #64685: (Linker error because `-wL as-needed` conflicts with
+                // `-C link-dead-code`). As described by this issue, the "link dead code" option
+                // produces incorrect binaries when compiled and linked under MSVC. The resulting
+                // Rust programs typically crash with a segmentation fault, or produce an empty
+                // "*.profraw" file (profiling counter results normally generated during program
+                // exit).
```

Comment:
I'm not sure if that issue is the right one or not, we don't seem to specify anything like `--as-needed` for the MSVC linker. (Though that may be the default behavior.) We might want to open a new one.

Style nit: when possible, try to keep summaries of an issue to one or two lines and refer to the issue itself for further details.

Reply:
No, you're probably right that the title isn't an exact match for the specific issue we have with InstrProf counters and Rust's `--link-dead-code` option.

But on the other hand, it sure is suspicious and highly coincidental that `--link-dead-code` is causing problems in other profiling solutions.

They aren't exactly the same problems, and not the same linkers, I agree. That alone may be confusing enough to justify splitting them out.

But I think they should be cross referenced at least.

I'll make a new issue and update the Issue references (and trim out some of the detail in the code comments).

BTW, since our issue related to MSVC specifically, then yes it's likely not related to `--as-needed` at all. I suspect (because of the leading double-dash) this is not an MSVC `LINK.EXE` option. But Rust does appear to enable `--as-needed` in Gcc:

https://github.com/rust-lang/rust/blob/9c9db3ab4b5fcb3349ec744c4f168dd3e753d57b/src/librustc_target/spec/linux_base.rs#L8-L15

Reply:
Filed a new issue and updated the reference

Reply:
A new issue with a reference between them is perfect, thanks!